pkg/cli/admin/release/git.go

```diff
@@ -97,7 +98,13 @@ func (g *git) VerifyCommit(repo, commit string) (bool, error) {
 	return err == nil, nil
 }
 
-func (g *git) CheckoutCommit(repo, commit string) error {
+func (g *git) CheckoutCommit(repo, commit string, out, errOut io.Writer) error {
+	// to reduce size requirements, clones are normally bare; to checkout a git commit, we must convert it to a normal
+	// git directory
+	if err := g.ensureFullClone(out, errOut); err != nil {
```

Comment:
If we have to perform full clone eventually, how do we reduce the disk usage size?. It seems to me that we are reducing disk size insignificant comparing to the complexity we are adding according to the latest status of this PR.

What was the problem in Github API side, rate limit errors from Github, different outputs, etc.?

Reply:
The `CheckoutCommit` command is only executed if a user runs the `oc adm release extract --git=...` command, and `CheckoutCommit` is the only function using `ensureFullClone`. It does not get executed by `oc release adm info --changelog/--bugs`, so the changelog generation remains unchanged, only requiring ~1GB of space.

The problem on the Github API side was speed. Running it locally with a github token would take about 5 minutes, and even running in the app.ci cluster using Prow's Github API cache, the changelog generation still took 2 minutes due to to the speed at which the API calls were returned (uncached in app.ci took ~3 minutes). I tested running all the API calls in parallel as well, but that only reduced it to ~1.5 minutes due to rate limiting. If you don't use an access token, you would also immediately hit the hourly rate limit for changelogs that weren't small nightly changes. With this PR, the generation of the first changelog takes about 2-3 minutes locally and 1 minute in the app.ci cluster. Future runs also only take a couple seconds, whereas the github method doesn't improve for future runs.